File: Utils/Position.py
import math

from Utils.Direction import Direction


class Position:
    def __init__(self, row, column):
        self.r = row
        self.c = column

    def neighbors(self, mode='orthogonal') -> list['Position']:
        neighbors = [self.down, self.up, self.right, self.left]
        if mode == 'diagonal':
            neighbors.extend([self.up_left, self.up_right, self.down_left, self.down_right])
        return neighbors

    def straddled_neighbors(self) -> set['Position']:
        r_floor = math.floor(self.r)
        c_floor = math.floor(self.c)
        r_ceil = math.ceil(self.r)
        c_ceil = math.ceil(self.c)
        return {Position(r_floor, c_floor), Position(r_floor, c_ceil), Position(r_ceil, c_floor), Position(r_ceil, c_ceil)}

    def direction_to(self, other: 'Position') -> Direction:
        if other is None or self == other:
            return Direction(Direction.none())
        if self.r == other.r:
            if self.c < other.c:
                return Direction.right()
            return Direction.left()
        if self.c == other.c:
            if self.r < other.r:
                return Direction.down()
            return Direction.up()
        return Direction(Direction.none())

    def direction_from(self, other: 'Position') -> Direction:
        return other.direction_to(self)

    def distance_to(self, other: 'Position') -> float:
        return math.sqrt(math.pow(abs(self.r - other.r), 2) + math.pow(abs(self.c - other.c), 2))

    def next(self, direction: Direction, count=1) -> 'Position':
        if direction == Direction.down():
            return Position(self.r + count, self.c)
        if direction == Direction.right():
            return Position(self.r, self.c + count)
        if direction == Direction.up():
            return Position(self.r - count, self.c)
        if direction == Direction.left():
            return Position(self.r, self.c - count)
        return self

    @property
    def left(self):
        return Position(self.r, self.c - 1)

    @property
    def right(self):
        return Position(self.r, self.c + 1)

    @property
    def up(self):
        return Position(self.r - 1, self.c)

    @property
    def down(self):
        return Position(self.r + 1, self.c)

    @property
    def up_left(self):
        return Position(self.r - 1, self.c - 1)

    @property
    def up_right(self):
        return Position(self.r - 1, self.c + 1)

    @property
    def down_left(self):
        return Position(self.r + 1, self.c - 1)

    @property
    def down_right(self):
        return Position(self.r + 1, self.c + 1)

    def get_positions_to(self, position: 'Position') -> list['Position']:
        if self.r == position.r:
            return [Position(self.r, c) for c in range(min(self.c, position.c) + 1, max(self.c, position.c))]
        if self.c == position.c:
            return [Position(r, self.c) for r in range(min(self.r, position.r) + 1, max(self.r, position.r))]
        return []

    def symmetric(self, position, to_int=True) -> 'Position':
        if to_int:
            return Position(int(2 * position.r - self.r), int(2 * position.c - self.c))
        return Position(2 * position.r - self.r, 2 * position.c - self.c)

    def __eq__(self, other):
        return isinstance(other, Position) and self.r == other.r and self.c == other.c

    def __hash__(self):
        return hash((self.r, self.c))

    def __str__(self):
        return f'({self.r}, {self.c})'

    def __repr__(self):
        return f'Position{self.__str__()}'

    def __add__(self, other):
        return Position(self.r + other.r, self.c + other.c)

    def __sub__(self, other):
        return Position(self.r - other.r, self.c - other.c)

    def __mul__(self, other):
        return Position(self.r * other, self.c * other)

    def __truediv__(self, other):
        return Position(self.r / other, self.c / other)

    def __floordiv__(self, other):
        return Position(self.r // other, self.c // other)

    def __mod__(self, other):
        return Position(self.r % other, self.c % other)

    def __lt__(self, other):
        return self.r < other.r or (self.r == other.r and self.c < other.c)

    def __le__(self, other):
        return self.r <= other.r or (self.r == other.r and self.c <= other.c)

    def __gt__(self, other):
        return self.r > other.r or (self.r == other.r and self.c > other.c)

    def __ge__(self, other):
        return self.r >= other.r or (self.r == other.r and self.c >= other.c)

    def __getitem__(self, item):
        return self.r if item == 0 else self.c

    def __setitem__(self, key, value):
        if key == 0:
            self.r = value
        else:
            self.c = value

    def __iter__(self):
        return iter([self.r, self.c])

    def __neg__(self):
        return Position(-self.r, -self.c)

File: KonarupuSolverCSharp/Position.cs
using System;
using System.Collections.Generic;

namespace KonarupuSolverCSharp
{
    public class Position
    {
        public int R { get; set; }
        public int C { get; set; }

        public Position(int row, int column)
        {
            R = row;
            C = column;
        }

        public List<Position> Neighbors(string mode = "orthogonal")
        {
            if (mode == "orthogonal")
            {
                return new List<Position> { Up, Left, Down, Right };
            }
            if (mode == "diagonal")
            {
                return new List<Position> { Up, UpLeft, Left, DownLeft, Down, DownRight, Right, UpRight };
            }
            throw new ArgumentException($"Unknown mode {mode}");
        }

        public HashSet<Position> StraddledNeighbors()
        {
            int rFloor = (int)Math.Floor((double)R);
            int cFloor = (int)Math.Floor((double)C);
            int rCeil = (int)Math.Ceiling((double)R);
            int cCeil = (int)Math.Ceiling((double)C);
            return new HashSet<Position>
            {
                new Position(rFloor, cFloor),
                new Position(rFloor, cCeil),
                new Position(rCeil, cFloor),
                new Position(rCeil, cCeil)
            };
        }

        public Direction DirectionTo(Position other)
        {
            if (other == null || this.Equals(other))
            {
                return Direction.None;
            }
            if (R == other.R)
            {
                return C < other.C ? Direction.Right : Direction.Left;
            }
            if (C == other.C)
            {
                return R < other.R ? Direction.Down : Direction.Up;
            }
            return Direction.None;
        }

        public Direction DirectionFrom(Position other)
        {
            return other.DirectionTo(this);
        }

        public double DistanceTo(Position other)
        {
            if (R == other.R)
            {
                return Math.Abs(C - other.C);
            }
            if (C == other.C)
            {
                return Math.Abs(R - other.R);
            }
            return Math.Sqrt(Math.Pow(R - other.R, 2) + Math.Pow(C - other.C, 2));
        }

        public Position After(Direction direction, int count = 1)
        {
            if (direction == Direction.Down) return new Position(R + count, C);
            if (direction == Direction.Right) return new Position(R, C + count);
            if (direction == Direction.Up) return new Position(R - count, C);
            if (direction == Direction.Left) return new Position(R, C - count);
            return this;
        }

        public Position Before(Direction direction, int count = 1)
        {
            return After(direction.Opposite(), count);
        }

        public Position Left => new Position(R, C - 1);
        public Position Right => new Position(R, C + 1);
        public Position Up => new Position(R - 1, C);
        public Position Down => new Position(R + 1, C);
        public Position UpLeft => new Position(R - 1, C - 1);
        public Position UpRight => new Position(R - 1, C + 1);
        public Position DownLeft => new Position(R + 1, C - 1);
        public Position DownRight => new Position(R + 1, C + 1);

        public List<Position> AllPositionsBetween(Position position)
        {
            var positions = new List<Position>();
            if (R == position.R)
            {
                for (int c = Math.Min(C, position.C) + 1; c < Math.Max(C, position.C); c++)
                {
                    positions.Add(new Position(R, c));
                }
            }
            else if (C == position.C)
            {
                for (int r = Math.Min(R, position.R) + 1; r < Math.Max(R, position.R); r++)
                {
                    positions.Add(new Position(r, C));
                }
            }
            return positions;
        }

        public List<Position> AllPositionsAndBoundsBetween(Position position)
        {
            var positions = new List<Position>();
            if (R == position.R)
            {
                for (int c = Math.Min(C, position.C); c <= Math.Max(C, position.C); c++)
                {
                    positions.Add(new Position(R, c));
                }
            }
            else if (C == position.C)
            {
                for (int r = Math.Min(R, position.R); r <= Math.Max(R, position.R); r++)
                {
                    positions.Add(new Position(r, C));
                }
            }
            return positions;
        }

        public Position Symmetric(Position position, bool toInt = true)
        {
            if (toInt)
            {
                return new Position((int)(2 * position.R - this.R), (int)(2 * position.C - this.C));
            }
            return new Position(2 * position.R - this.R, 2 * position.C - this.C);
        }

        public bool IsOnRow()
        {
            return R == Math.Floor((double)R);
        }

        public bool IsOnColumn()
        {
            return C == Math.Floor((double)C);
        }

        public override bool Equals(object obj)
        {
            return obj is Position pos && R == pos.R && C == pos.C;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(R, C);
        }

        public override string ToString()
        {
            return $"({R}, {C})";
        }

        public static Position operator +(Position a, Position b) => new Position(a.R + b.R, a.C + b.C);
        public static Position operator -(Position a, Position b) => new Position(a.R - b.R, a.C - b.C);
        public static Position operator *(Position a, int b) => new Position(a.R * b, a.C * b);
        public static Position operator /(Position a, int b) => new Position(a.R / b, a.C / b);
    }
}